graphql_compiler/macros/macro_expansion.py

```diff
@@ -28,21 +29,144 @@ def _merge_non_overlapping_dicts(merge_target, new_data):
     return result
 
 
-def _expand_specific_macro_edge(macro_definition_selection_set, selection_ast, subclass_sets=None):
+def _merge_selection_sets(selection_set_a, selection_set_b):
+    """Merge selection sets, merging directives on name conflict."""
+    if selection_set_a is None:
+        return selection_set_b
+    if selection_set_b is None:
+        return selection_set_a
+
+    # Convert to dict
+    selection_dict_a = get_uniquely_named_objects_by_name(selection_set_a.selections)
+    selection_dict_b = get_uniquely_named_objects_by_name(selection_set_b.selections)
+
+    # Compute intersection by name
+    common_selection_dict = dict()
+    common_fields = set(selection_dict_a.keys()) & set(selection_dict_b.keys())
+    for field_name in common_fields:
+        field_a = selection_dict_a[field_name]
+        field_b = selection_dict_b[field_name]
+        if field_a.selection_set is not None or field_b.selection_set is not None:
+            raise GraphQLCompilationError('TODO')
+
+        merged_field = deepcopy(field_a)
+        merged_field.directives += field_b.directives
+        common_selection_dict[field_name] = merged_field
+
+    # Merge dicts
+    merged_selection_dict = copy(selection_dict_a)
+    merged_selection_dict.update(selection_dict_b)
+    merged_selection_dict.update(common_selection_dict)
+
+    # Remove pro-forma fields if allowed
+    if len(merged_selection_dict) > 1:
+        merged_selection_dict = {
+            name: ast
+            for name, ast in six.iteritems(merged_selection_dict)
+            if ast.selection_set is not None or len(ast.directives) > 0
+        }
+
+    # Get a deterministic ordering of the merged selections
+    selection_name_order = [
+        ast.name.value
+        for ast in selection_set_a.selections
+        if ast.name.value not in selection_dict_b
+    ] + [
+        ast.name.value
+        for ast in selection_set_b.selections
+    ]
+
+    # Make sure that all property fields come before all vertex fields.
+    return SelectionSet([
+        merged_selection_dict[name]
+        for name in selection_name_order
+        if name in merged_selection_dict and merged_selection_dict[name].selection_set is None
+    ] + [
+        merged_selection_dict[name]
+        for name in selection_name_order
+        if name in merged_selection_dict and merged_selection_dict[name].selection_set is not None
+    ])
+
+
+def _merge_selection_into_target(target_ast, selection_ast, subclass_sets=None):
+    """Add the selections, directives, and coercions from the selection_ast to the target_ast."""
+    # Deal with coercions at the target
+    continuation_ast = selection_ast
+    first_selection = selection_ast.selection_set.selections[0]
```

Comment:
How do we know that `selection_set` is not `None` here? Also, the remainder of `selection_ast.selection_set.selections` is apparently never processed, is that okay? It seems suspicious.

Reply:
Good catch. I was making the assumption that if there's an inline fragment, it's the first selection and there's no more selections.